### PR TITLE
Implement auto-merge bot functionality

### PR DESCRIPTION
**Pull request by mbestavros**
This code implements auto-merge for the Enarx organization bot.

While functionally complete, I'm submitting in draft form until I know the name of the encrypted secret to use (currently `secrets.BOT_TOKEN`).

This code should run every 15 minutes, and will leave logs about which PRs were and were not auto-merged.

### CODE REVIEW

File: .github/workflows/scheduled-run.yml [hunk 1]
```diff
@@ -0,0 +1,16 @@
+name: scheduled-run
+
+on:
+  schedule:
+    - cron:  '*/15 * * * *'
+
+jobs:
+  scheduled-run:
+    runs-on: ubuntu-latest
+    container: quay.io/enarx/fedora
```
Comment (npmccallum) [inline .github/workflows/scheduled-run.yml:10]: We probably don't want to use our container for this. It is rather large.

File: .github/workflows/scheduled-run.yml [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+    steps:
+    - uses: actions/checkout@v2
+    - name: scheduled-run
+      env:
+        BOT_TOKEN: ${{ secrets.BOT_TOKEN }}
+      run: scripts/exec $BOT_TOKEN "enarx" $GITHUB_EVENT_PATH
```
Comment (npmccallum) [inline .github/workflows/scheduled-run.yml:16]:
```suggestion
      run: scripts/exec $BOT_TOKEN "enarx"
```

File: scripts/exec [hunk 3]
```diff
@@ -0,0 +1,41 @@
+#!/usr/bin/python3
+# SPDX-License-Identifier: Apache-2.0
+
+import sys
+from github import Github
+
+# Get inputs from shell
+(token, organization_name, path) = sys.argv[1:4]
```
Comment (npmccallum) [inline scripts/exec:8]:
```suggestion
(token, organization_name) = sys.argv[1:4]
```

File: scripts/exec [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+for repo in org_repos:
+    print("Organizing " + repo.name)
+
+    # Automatically merge PRs that meet criteria
+    prs = repo.get_pulls()
+    for pr in prs:
+        pr_reviews = pr.get_reviews()
+        
+        # Parse reviews for merge criteria.
+        changes_requested = False
```
Comment (npmccallum) [inline scripts/exec:23]:
```suggestion
        changes_requested = 0
```

File: scripts/exec [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+        approving_reviews = 0
+        for review in pr_reviews:
+            if review.state == "APPROVED":
+                approving_reviews += 1
+            if review.state == "CHANGES_REQUESTED":
+                changes_requested = True
```
Comment (npmccallum) [inline scripts/exec:29]:
```suggestion
                changes_requested += 1
```

File: scripts/exec [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+        
+        # If all merge criteria pass, merge the PR.
+        if pr.mergeable == True and approving_reviews >= 2 and changes_requested == False:
```
Comment (npmccallum) [inline scripts/exec:32]:
```suggestion
        if pr.mergeable == True and approving_reviews >= 2 and changes_requested == 0:
```

File: scripts/exec [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+    # Automatically merge PRs that meet criteria
+    prs = repo.get_pulls()
+    for pr in prs:
+        pr_reviews = pr.get_reviews()
+        
+        # Parse reviews for merge criteria.
+        changes_requested = False
+        approving_reviews = 0
+        for review in pr_reviews:
+            if review.state == "APPROVED":
```
Comment (npmccallum) [inline scripts/exec:26]: What other states are there? There could be subtle bugs here...

File: scripts/exec [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+                approving_reviews += 1
+            if review.state == "CHANGES_REQUESTED":
+                changes_requested = True
+        
+        # If all merge criteria pass, merge the PR.
+        if pr.mergeable == True and approving_reviews >= 2 and changes_requested == False:
+            print("Merging PR " + str(pr.number) + " from " + repo.name)
```
Comment (npmccallum) [inline scripts/exec:33]: For this, and all other strings, use a [format specifier](https://pyformat.info) or [f-string](https://realpython.com/python-f-strings/) (preferred).

File: .github/workflows/scheduled-run.yml (hunk 1 above)
Comment (mbestavros) [reply .github/workflows/scheduled-run.yml:10]: I suppose the only things we really need are Python and PyGithub. Makes sense to have a container with only those things included.

File: scripts/exec (hunk 7 above)
Comment (mbestavros) [reply scripts/exec:26]: There is surprisingly no documentation on the `state` field for `PullRequestReview` on Github's end: https://developer.github.com/v3/pulls/reviews/

They do provide examples in the sample responses, but no definitive listing of all possible states. The ones I can see are:

- APPROVED
- CHANGES_REQUESTED
- PENDING
- DISMISSED

The last one might cause some problems with our criteria that all "Changes Requested" reviews be updated to "Approved", though I don't know exactly how that state works.

File: .github/workflows/scheduled-run.yml (hunk 1 above)
Comment (npmccallum) [reply .github/workflows/scheduled-run.yml:10]: If `ubuntu-latest` has `PyGithub`, we probably don't need a container at all.

Comment (mbestavros) [reply .github/workflows/scheduled-run.yml:10]: Very good point. At worst, we'd probably have to `pip install` it, which would still take less time than spinning up a container would.

**Comment by mbestavros** [comment]
Closing in favor of https://github.com/enarxbot/enarxbot/pull/2. I committed the cardinal sin of making a PR from master...

@npmccallum I've addressed all the minor feedback from your first review, and moved the larger points over to the new PR.

**Pull request closed without merging**